src/sub.c

```diff
@@ -23,54 +23,99 @@
 #include "util.h"
 #include "js.h"
 #include "opts.h"
+#include "glib/glib.h"
 
 #ifdef DEV_MODE
 
-static void _retain(natsSubscription *sub)   { sub->refs++; }
-static void _release(natsSubscription *sub)  { sub->refs--; }
-
-void natsSub_Lock(natsSubscription *sub)     { natsMutex_Lock(sub->mu);   }
-void natsSub_Unlock(natsSubscription *sub)   { natsMutex_Unlock(sub->mu); }
+static inline int _retain(natsSubscription *sub) { return ++(sub->refs); }
+static inline int _release(natsSubscription *sub) { return --(sub->refs); }
 
 #else
 
-#define _retain(s)  ((s)->refs++)
-#define _release(s) ((s)->refs--)
+#define _retain(s) (++((s)->refs))
+#define _release(s) (--((s)->refs))
 
 #endif // DEV_MODE
 
-#define SUB_DLV_WORKER_LOCK(s)      if ((s)->libDlvWorker != NULL) \
-                                        natsMutex_Lock((s)->libDlvWorker->lock)
-
-#define SUB_DLV_WORKER_UNLOCK(s)    if ((s)->libDlvWorker != NULL) \
-                                        natsMutex_Unlock((s)->libDlvWorker->lock)
-
 bool testDrainAutoUnsubRace = false;
 
-static void
-_freeSubscription(natsSubscription *sub)
+static inline void _destroyControlMessage(natsMsg *msg)
 {
-    natsMsg *m;
+    if (msg != NULL)
+    {
+        natsMsg_clearNoDestroy(msg);
+        natsMsg_Destroy(msg);
+    }
+}
 
-    if (sub == NULL)
+static inline void _freeControlMessages(natsSubscription *sub)
+{
+    if (sub->control == NULL)
         return;
 
-    while ((m = sub->msgList.head) != NULL)
+    _destroyControlMessage(sub->control->sub.timeout);
+    _destroyControlMessage(sub->control->sub.close);
+    _destroyControlMessage(sub->control->sub.drain);
+    _destroyControlMessage(sub->control->batch.expired);
+    _destroyControlMessage(sub->control->batch.missedHeartbeat);
+    NATS_FREE(sub->control);
+}
+
+static inline natsStatus _createControlMessage(natsMsg **msg, natsSubscription *sub)
+{
+    natsStatus s = natsMsg_create(msg, NULL, 0, NULL, 0, NULL, 0, -1);
+    if (s == NATS_OK)
     {
-        sub->msgList.head = m->next;
-        natsMsg_Destroy(m);
+        natsMsg_setNoDestroy(*msg);
+        (*msg)->sub = sub;
     }
+    return s;
+}
 
-    NATS_FREE(sub->subject);
-    NATS_FREE(sub->queue);
+// Sets up the default (user-thread NextMsg-only) dispatcher for a sub. Avoid
+// the use of term "pull" since it's not to be confused with JetStream consumer
+// pull requests.
+static natsStatus
+_initOwnDispatcher(natsSubscription *sub)
+{
+    natsStatus s = NATS_OK;
+
+    if (sub->ownDispatcher.dedicatedTo != NULL)
+        return NATS_ILLEGAL_STATE; // already started
 
-    if (sub->deliverMsgsThread != NULL)
+    sub->ownDispatcher.dedicatedTo = sub;
+    sub->ownDispatcher.mu = sub->mu;
+    s = natsCondition_Create(&sub->ownDispatcher.cond);
+    return s;
+}
+
+static inline void _cleanupOwnDispatcher(natsSubscription *sub)
+{
+    nats_destroyQueuedMessages(&sub->ownDispatcher.queue);
```

Comment:
Again, cleanup/destroy functions may be called as part of the creation of the object and you have to assume that the failure could happen at any point and therefore some fields may not have been initialized. So here ownDispatcher seem to be a struct, so it won't be `NULL`, but `queue` may, yet `nats_destroyQueuedMessages` does not seem to be checking for `NULL` there.

Reply:
Same, `queue` is not a pointer; a NULL-ed out struct, so the for loop bails.